Matlab wrapper uses python 3

Problem:
[ 56%] Generating wrap/gtsam/gtsam_wrapper.cpp
Ignoring classes: ['gtsam::Point2', 'gtsam::Point3']
Traceback (most recent call last):
  File "/path/gtsam/wrap/matlab_wrapper.py", line 1770, in <module>
    cc_content = wrapper.wrap()
  File "/path/gtsam/wrap/matlab_wrapper.py", line 1657, in wrap
    self.wrap_namespace(self.module)
  File "/path/gtsam/wrap/matlab_wrapper.py", line 1096, in wrap_namespace
    self.wrap_namespace(element, namespaces)
  File "/path/gtsam/wrap/matlab_wrapper.py", line 1101, in wrap_namespace
    class_text = self.wrap_instantiated_class(element, "".join(namespace.full_namespaces()))
  File "/path/gtsam/wrap/matlab_wrapper.py", line 1037, in wrap_instantiated_class
    ).splitlines()) + '\n'
  File "/path/gtsam/wrap/matlab_wrapper.py", line 698, in wrap_class_constructors
    methods_wrap = textwrap.indent(textwrap.dedent("""\

To generate the Matlab wrapper it requires python 3.  I fixed by changing "Default" to my python 3 version.

https://github.com/borglab/gtsam/blob/de75367e5e8f49f468474a75e9fd4bdb15fc856e/CMakeLists.txt#L104

My apologies that it cut off! First error encountered was:

------------------------------------------------------------------------------------------------------

[ 56%] Generating wrap/gtsam/gtsam_wrapper.cpp
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 87
    print(message, file=sys.stderr)
                       ^
SyntaxError: invalid syntax
gtsam/CMakeFiles/gtsam_matlab_wrapper.dir/build.make:76: recipe for target 'gtsam/wrap/gtsam/gtsam_wrapper.cpp' failed
make[2]: *** [gtsam/wrap/gtsam/gtsam_wrapper.cpp] Error 1
CMakeFiles/Makefile2:1188: recipe for target 'gtsam/CMakeFiles/gtsam_matlab_wrapper.dir/all' failed
make[1]: *** [gtsam/CMakeFiles/gtsam_matlab_wrapper.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

------------------------------------------------------------------------------------------------------

This should have been my warning but I used 
`from __future__ import print_function
`
to bypass that. The next error was:

------------------------------------------------------------------------------------------------------

[ 58%] Building CXX object examples/CMakeFiles/Pose3SLAMExampleExpressions_BearingRangeWithTransform.dir/Pose3SLAMExampleExpressions_BearingRangeWithTransform.cpp.o
Ignoring classes: ['gtsam::Point2', 'gtsam::Point3']
Traceback (most recent call last):
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 1769, in <module>
    cc_content = wrapper.wrap()
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 1656, in wrap
    self.wrap_namespace(self.module)
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 1095, in wrap_namespace
    self.wrap_namespace(element, namespaces)
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 1100, in wrap_namespace
    class_text = self.wrap_instantiated_class(element, "".join(namespace.full_namespaces()))
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 1036, in wrap_instantiated_class
    ).splitlines()) + '\n'
  File "/home/mark/installs/gtsam/wrap/matlab_wrapper.py", line 697, in wrap_class_constructors
    methods_wrap = textwrap.indent(textwrap.dedent("""\
AttributeError: 'module' object has no attribute 'indent'
gtsam/CMakeFiles/gtsam_matlab_wrapper.dir/build.make:76: recipe for target 'gtsam/wrap/gtsam/gtsam_wrapper.cpp' failed
make[2]: *** [gtsam/wrap/gtsam/gtsam_wrapper.cpp] Error 1
CMakeFiles/Makefile2:1188: recipe for target 'gtsam/CMakeFiles/gtsam_matlab_wrapper.dir/all' failed
make[1]: *** [gtsam/CMakeFiles/gtsam_matlab_wrapper.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

------------------------------------------------------------------------------------------------------

I wasn't suggesting for it be made Python 2 compatible but rather that there could be a catch for this to ensure Python 3 is used. In my case, the default was still Python 2.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -127,7 +127,11 @@ if(GTSAM_BUILD_PYTHON)
     endif()
 
     if(${GTSAM_PYTHON_VERSION} STREQUAL "Default")
-        set(GTSAM_PYTHON_VERSION "${Python3_VERSION_MAJOR}.${Python3_VERSION_MINOR}" CACHE STRING "The version of Python to build the wrappers against." FORCE)
+        set(GTSAM_PYTHON_VERSION "${Python3_VERSION_MAJOR}.${Python3_VERSION_MINOR}"
+                CACHE
+                STRING
+                "The version of Python to build the wrappers against."
+                FORCE)
     endif()
 
     if(GTSAM_UNSTABLE_BUILD_PYTHON)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -14,6 +14,11 @@ set (GTSAM_VERSION_PATCH 0)
 math (EXPR GTSAM_VERSION_NUMERIC "10000 * ${GTSAM_VERSION_MAJOR} + 100 * ${GTSAM_VERSION_MINOR} + ${GTSAM_VERSION_PATCH}")
 set (GTSAM_VERSION_STRING "${GTSAM_VERSION_MAJOR}.${GTSAM_VERSION_MINOR}.${GTSAM_VERSION_PATCH}")
 
+set (CMAKE_PROJECT_VERSION ${GTSAM_VERSION_STRING})
+set (CMAKE_PROJECT_VERSION_MAJOR ${GTSAM_VERSION_MAJOR})
+set (CMAKE_PROJECT_VERSION_MINOR ${GTSAM_VERSION_MINOR})
+set (CMAKE_PROJECT_VERSION_PATCH ${GTSAM_VERSION_PATCH})
+
 ###############################################################################
 # Gather information, perform checks, set defaults
 
@@ -113,6 +118,18 @@ if(GTSAM_INSTALL_MATLAB_TOOLBOX AND NOT BUILD_SHARED_LIBS)
 endif()
 
 if(GTSAM_BUILD_PYTHON)
+    # Get info about the Python3 interpreter
+    # https://cmake.org/cmake/help/latest/module/FindPython3.html#module:FindPython3
+    find_package(Python3 COMPONENTS Interpreter Development)
+
+    if(NOT ${Python3_FOUND})
+        message(FATAL_ERROR "Cannot find Python3 interpreter. Please install Python >= 3.6.")
+    endif()
+
+    if(${GTSAM_PYTHON_VERSION} STREQUAL "Default")
+        set(GTSAM_PYTHON_VERSION "${Python3_VERSION_MAJOR}.${Python3_VERSION_MINOR}" CACHE STRING "The version of Python to build the wrappers against." FORCE)
+    endif()
+
     if(GTSAM_UNSTABLE_BUILD_PYTHON)
         if (NOT GTSAM_BUILD_UNSTABLE)
             message(WARNING "GTSAM_UNSTABLE_BUILD_PYTHON requires the unstable module to be enabled.")

--- a/cmake/FindNumPy.cmake
+++ b/cmake/FindNumPy.cmake
@@ -40,17 +40,9 @@
 
 # Finding NumPy involves calling the Python interpreter
 if(NumPy_FIND_REQUIRED)
-  if(GTSAM_PYTHON_VERSION STREQUAL "Default")
-    find_package(PythonInterp REQUIRED)
-  else()
-      find_package(PythonInterp ${GTSAM_PYTHON_VERSION} EXACT REQUIRED)
-  endif()
+  find_package(PythonInterp ${GTSAM_PYTHON_VERSION} EXACT REQUIRED)
 else()
-  if(GTSAM_PYTHON_VERSION STREQUAL "Default")
-    find_package(PythonInterp)
-  else()
-    find_package(PythonInterp ${GTSAM_PYTHON_VERSION} EXACT)
-  endif()
+  find_package(PythonInterp ${GTSAM_PYTHON_VERSION} EXACT)
 endif()
 
 if(NOT PYTHONINTERP_FOUND)

--- a/cmake/GtsamMatlabWrap.cmake
+++ b/cmake/GtsamMatlabWrap.cmake
@@ -215,19 +215,15 @@ function(wrap_library_internal interfaceHeader linkLibraries extraIncludeDirs ex
 	# Set up generation of module source file
 	file(MAKE_DIRECTORY "${generated_files_path}")
 
-	if(GTSAM_PYTHON_VERSION STREQUAL "Default")
-		find_package(PythonInterp REQUIRED)
-		find_package(PythonLibs REQUIRED)
-	else()
-		find_package(PythonInterp
-				${GTSAM_PYTHON_VERSION}
-				EXACT
-				REQUIRED)
-		find_package(PythonLibs
-				${GTSAM_PYTHON_VERSION}
-				EXACT
-				REQUIRED)
-	endif()
+    find_package(PythonInterp
+            ${GTSAM_PYTHON_VERSION}
+            EXACT
+            REQUIRED)
+    find_package(PythonLibs
+            ${GTSAM_PYTHON_VERSION}
+            EXACT
+            REQUIRED)
+
 
 	set(_ignore gtsam::Point2
 			gtsam::Point3)

